corehq/apps/userreports/views.py

```diff
@@ -439,32 +443,97 @@ def preview_data_source(request, domain, config_id):
+    start = end - datetime.timedelta(days=int(value))
+    return column.between(start, end)
+
+
+def _range_filter(column, value):
+    try:
+        start, end = value.split('..')
+    except ValueError:
+        raise UserQueryError(_('Ranges must have the format "start..end"'))
+    return column.between(start, end)
+
+
+sql_directives = [
+    # (suffix matching url parameter, callable returning a filter),
+    ('-lastndays', _last_n_days),
+    ('-range', _range_filter),
+]
+
+
+def process_url_params(params, columns):
+    """
+    Converts a dictionary of parameters from the user to sql filters.
+
+    If a parameter is of the form <field name>-<suffix>, where suffix is
+    defined in `sql_directives`, the corresponding function is used to
+    produce a filter.
+    """
+    format_ = params.get('$format', Format.UNZIPPED_CSV)
+    keyword_filters = {}
+    sql_filters = []
+    for key, value in params.items():
+        if key == '$format':
+            continue
+
+        for suffix, fn in sql_directives:
+            if key.endswith(suffix):
+                field = key.rstrip(suffix)
+                if field not in columns:
+                    raise UserQueryError(_('No field named {}').format(field))
+                sql_filters.append(fn(columns[field], value))
+                break
+        else:
+            if key in columns:
+                keyword_filters[key] = value
+            else:
+                raise UserQueryError(_('Invalid filter parameter: {}')
+                                     .format(key))
+    return ExportParameters(format_, keyword_filters, sql_filters)
+
+
 @login_or_basic
 @require_permission(Permissions.view_reports)
 def export_data_source(request, domain, config_id):
-    format = request.GET.get('format', Format.UNZIPPED_CSV)
     config = get_document_or_404(DataSourceConfiguration, domain, config_id)
     table = get_indicator_table(config)
     q = Session.query(table)
-    column_headers = [col['name'] for col in q.column_descriptions]
 
-    # apply filtering if any
-    filter_values = {key: value for key, value in request.GET.items() if key != 'format'}
-    for key in filter_values:
-        if key not in column_headers:
-            return HttpResponse('Invalid filter parameter: {}'.format(key), status=400)
-    q = q.filter_by(**filter_values)
+    try:
+        params = process_url_params(request.GET, table.columns)
+    except UserQueryError as e:
+        return HttpResponse(e.message, status=400)
+
+    q = q.filter_by(**params.keyword_filters)
+    for sql_filter in params.sql_filters:
+        q = q.filter(sql_filter)
 
     # build export
     def get_table(q):
-        yield column_headers
+        yield table.columns.keys()
         for row in q:
             yield row
 
+    try:
+        full_table = list(get_table(q))
+    except exc.DataError:
+        msg = _("There was a problem executing your query, please make sure "
+                "your parameters are valid.")
```

Comment:
I'm undecided on whether or not to include `DataError.message` in the response.  It'd certainly be helpful ("parameter opened_on_1231231 something something .. is the wrong type"), but I'm not sure whether it's a good idea to blindly pass through error messages - could there be sensitive information in there?